<b>Enoncé :</b>

Soit le stock d'une entrprise de vente de produits :

            label           puht        quantité
            télé            1201.988    5    
            réfrégirateur   2100.744    3    
            micro-onde      350.133     15 

L'objectif de ce travail est d'étudier ce stock en utilisant les deux concepts du langage Python : `List comprehension` et `Fonctions`.

Vous allez suivre les étapes de la démarche suivante pour réaliser cette étude.

<b>Etape 1 : Collecte des données</b>

<u>Question 1:</u>

Un produit est défini par :
- son nom (chaine de caractères) 
- son prix unitaire hors taxe (puht) (réel) 
- sa quantité (entier)

Dans cette étude, ces informations des produits sont stockés dans les 3 listes suivantes : 
- noms : elle va contenir la liste des noms
- puhts : elle va contenir la liste des prix unitaires hors taxe.
- quantites : elle va contenir les quantités des produits en stock.

Déclarer les listes `noms`, `puhts` et `quantites` comme vides.

In [66]:
#?

<u>Question 2 :</u>

Les informations des produits sont saisies à partir du clavier.

Définir une fonction `saisir_produit` qui permet de saisir et retourner le nom, le puht et la quantité d'un seul produit.

In [68]:
#?

In [3]:
#test
nom, puht, quantite=saisir_produit()

In [9]:
print(f"nom={nom},puht={puht},quantité={quantite}")

nom=tele,puht=1200.556,quantité=5


<u>Question 3 :</u>

Les informations saisies des produits sont stockées dans les 3 listes.

Définir une fonction `ajouter_produit` qui permet d'ajouter le nom, le puht et la quantité d'un seul produit dans les 3 listes.

In [69]:
#?

In [11]:
# test
ajouter_produit('tele',1200.556,5)

In [12]:
#test
noms

['tele']

In [13]:
#test
puhts

[1200.556]

In [14]:
#test
quantites

[5]

<u>Question 4 :</u>

On veut maintenant saisir et ajouter les informations des produits, qui existent dans le stock.

Ecrire un code Python dans lequel :
1. Déclarer de nouveau les 3 listes `noms` , `puhts` et `quantites` comme vides.
2. Implémente une boucle dans laquelle à chaque itération:
    - Saisir les informations d'un produit.
    - Ajouter ces informatiosn saisies dans les 3 listes en utilisant les deux fonctions `saisir_produit` et `ajouter_produit`. 
    - Demander à l'utilisateur de saisir `stop` s'il veut arrêter la boucle.

In [73]:
#?

In [74]:
#test
noms

['télé', 'réfrigérateur', 'micro-onde']

In [75]:
#test
puhts

[1200.998, 2100.665, 350.112]

In [76]:
#test
quantites

[5, 3, 15]

<b>Etape 2 : Calcul des informations sur les produits</b>

<u>Question 1 :</u> 

Déterminer les abbréviations de tous les produits et les mettre dans une liste `labels` selon la logique suivante :

        if longueur de nom > 5 Alors 
                abrv est composé de 3 premières lettres en majuscule
        Sinon 
                abrv est composé de 2 premières lettres en majuscule

In [79]:
#?

In [80]:
#test
labels

['TÉ', 'RÉF', 'MIC']

 <u>Question 2 :</u>
 
 Le prix unitaire tout taxe compris d'un produit est calculé selon la formule :

            puttc = puht * (1+tva)

 tva prend souvent la valeur 10%

Définir la fonction `calculer_putcc` qui calcule et retourne le prix tout taxe compris (puttc) à partir de puht et tva.

In [81]:
#?

In [82]:
#test
puttc=calculer_puttc(1000)
puttc

1100.0

In [83]:
#test
puttc=calculer_puttc(1000,0.15)
puttc

1150.0

<u>Question 3 :</u>

Calculer les puttcs de tous les produits dans la liste  `puttcs` avec tva=10% en utilisant la fonction `calculer_puttc`.

In [84]:
#?

In [85]:
#test
puttcs

[1321.0978000000002, 2310.7315000000003, 385.12320000000005]

<u>Question 4:</u>

Le montant d'un produit est calculé selon la formule suivante :

                montant = puttc * quantité

Définir la fonction `calculer_montant` qui clcule et retourne le montant d'un produit étant donné son puttc et sa quantité.

In [86]:
#?

In [87]:
#test
montant=calculer_montant(1000,5)
montant

5000

<u>Question 5 :</u>

Calculer les montants de tous les produits dans la liste `montants` en utilisant la fonction `calculer_montant`.

In [88]:
#?

In [89]:
#test
montants

[6605.489000000001, 6932.194500000001, 5776.848000000001]

<b>Etape 3 : Affichage des données</b>


Définir la fonction `afficher_produits` qui affiche les informations de tous les produits sous la forme d'une table

Vous aurez normalement l'affichage suivant :

    label     puht      puttc     montant   quantité
    TÉ        1201.0    1321.1    6605.5    5    
    RÉF       2100.7    2310.7    6932.2    3    
    MIC       350.1     385.1     5776.8    15  

In [90]:
#?

In [91]:
#test
afficher_produits()

label     puht      puttc     montant   quantité
TÉ        1201.0    1321.1    6605.5    5    
RÉF       2100.7    2310.7    6932.2    3    
MIC       350.1     385.1     5776.8    15   


<b>Etape 4 : Fonctionalité de recherche</b>

<u>Question 1 :</u>

Définir la fonction `rechercher_produit` qui recherche un produit étant donné son nom et retourne deux valeurs :
- Vrai s'il exsite soinon Faux
- L'indice du produit trouvé s'il exsite Sinon `None`

In [40]:
#?

In [44]:
#test
trouve,indice_cible=rechercher_produit('micro-onde')
print(f"trouve={trouve},indice={indice_cible}")

trouve=True,indice=2


In [45]:
#test
trouve,indice_cible=rechercher_produit('micro')
print(f"trouve={trouve},indice={indice_cible}")

trouve=False,indice=None


<u>Question 2:</u>

Ecrire le code Python qui :
- saisit un nom du produit cible
- recherche un produit qui porte ce nom cible en utilisant la fonction `rechercher_produit`
- Afficher un message selon logique :

            Si le produit existe Alors
                    Afficher "Le produit abbreviation_produit n'existe pas dans le stock"
            Sinon
                    Afficher "Le produit nom_produit existe dans le stock
                              Son puht est puht_produit
                              Sa quantité est quantité_produit
                              Son montant est montant_produit"

In [96]:
#?

Le produit TÉ se trouve !
Son puht est 1201.0
Sa quantité est 5
Son montant est 6605.5


<b>Etape 4 : Calcul des informations sur le stock</u>

<u>Question 1 :</u>

Calculer le montant total, qui est égale à la somme des montants des produits.

In [ ]:
#?

In [97]:
#test
print(f"Le montant total est {montant_total:.1f}")

Le montant total est 19314.5


<u>Question 2 :</u>

Ecrire un code Python qui détermine le produit le plus rentable (ayant le montant le plus elevé).

In [ ]:
#?

In [99]:
#test
print(f"Le produit le plus rentable est {labels[indice_cible]}")
print(f"Son montant est {montant_max:.1f}")

Le produit le plus rentable est TÉ
Son montant est 6932.2


<u>Question 3 :</u>

Ecrire un code Python qui trie les informations des produits dans toutes les listes (noms,puhts,quantites,puttcs,montants) selon leurs montants

Ce tri est descendant (de produit ayant le plus elevé jusqu'à le plus faible)

<font color=red>Note:</font>

Utiliser le la méthode de tri que vous convient.

<font color=red>Astuce :</font>

En Python, on peut permuter les valeurs de deux variables x et y comme suit :

        x , y = y , x

In [100]:
#?

In [101]:
#test
afficher_produits()

label     puht      puttc     montant   quantité
TÉ        2100.7    2310.7    6932.2    3    
RÉF       1201.0    1321.1    6605.5    5    
MIC       350.1     385.1     5776.8    15   
